cvat/apps/engine/cache.py

```diff
@@ -54,74 +60,199 @@
         else:
-            # compare checksum
-            item_data = item[0].getbuffer() if isinstance(item[0], io.BytesIO) else item[0]
-            item_checksum = item[2] if len(item) == 3 else None
-            if item_checksum != self._get_checksum(item_data):
-                slogger.glob.info(f"Recreating cache item {key} due to checksum mismatch")
-                item = create_item()
+            try:
+                with get_rq_lock_for_job(queue, rq_id, blocking_timeout=self._LOCK_TIMEOUT):
+                    rq_job = queue.fetch_job(rq_id)
+
+                    if not rq_job:
+                        rq_job = queue.enqueue(
+                            self._drop_return_value,
+                            self._create_and_set_cache_item,
+                            key,
+                            create_callback,
+                            *args,
+                            cache_item_ttl=cache_item_ttl,
+                            **kwargs,
+                            job_id=rq_id,
+                            result_ttl=self._RQ_JOB_RESULT_TTL,
+                            failure_ttl=self._RQ_JOB_FAILURE_TTL,
+                        )
+            except LockError:
+                raise TimeoutError(f"Cannot acquire lock for {key}")
+
+            if self._wait_for_rq_job(rq_job):
+                item = self._get_cache_item(key)
+            else:
+                raise TimeoutError(f"Chunk processing takes too long {key}")
+
+        slogger.glob.info(f"Ending to prepare chunk: key {key}")
 
         return item
 
     def _delete_cache_item(self, key: str):
         try:
-            self._cache.delete(key)
+            self._cache().delete(key)
             slogger.glob.info(f"Removed chunk from the cache: key {key}")
         except pickle.UnpicklingError:
             slogger.glob.error(f"Failed to remove item from the cache: key {key}", exc_info=True)
 
     def _get_cache_item(self, key: str) -> Optional[_CacheItem]:
-        slogger.glob.info(f"Starting to get chunk from cache: key {key}")
         try:
-            item = self._cache.get(key)
+            item = self._cache().get(key)
         except pickle.UnpicklingError:
             slogger.glob.error(f"Unable to get item from cache: key {key}", exc_info=True)
             item = None
-        slogger.glob.info(f"Ending to get chunk from cache: key {key}, is_cached {bool(item)}")
+
+        if not item:
+            return None
+
+        item_data = item[0].getbuffer() if isinstance(item[0], io.BytesIO) else item[0]
+        item_checksum = item[2] if len(item) == 4 else None
+        if item_checksum != self._get_checksum(item_data):
+            slogger.glob.info(f"Cache item {key} checksum mismatch")
+            return None
 
         return item
 
-    def _has_key(self, key: str) -> bool:
-        return self._cache.has_key(key)
+    def _validate_cache_item_timestamp(self, item: _CacheItem, expected_timestamp: datetime):
+        if item:
```

Comment:
The argument annotation says `bool(item)` cannot be `False`, please check the logic.